File: code/run_csweep_extended.py
"""
Push-off-coefficient robustness, consistent with the corrected (extended, no-fold)
family. For each c in {0.80, 1.00, 1.04, 1.20}, the prescribed push-off is
P(s) = c * alpha * tan(alpha); the periodic branch is traced by seeded
continuation over s in [0.01, 0.80] (same method as run_extend_family.py).

Shows that for every c the branch extends smoothly to low speed with no fold encountered over
the computed range, and N_1/2 rises steeply toward the slow end — i.e. the low-speed
weakening is not an artefact of the specific coefficient 1.04. No claim is made
about the limit v -> 0: there is no analytic asymptotic derivation, so only the
trend over the computed range is reported.

Output: newest/figures/sens_pushoff_c.png (+ .tiff, .pdf)  (N_1/2 vs v, log axis, one curve per c)
        newest/data/supp_csweep.csv                        (the traced branches)
"""
import os
import numpy as np
import matplotlib.pyplot as plt
import revision_numerics as R

CS = [0.80, 1.00, 1.04, 1.20]
HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.abspath(os.path.join(HERE, ".."))
OUT = os.path.join(ROOT, "figures", "sens_pushoff_c.png")


def solve_at(q, c, seed):
    """q is the branch parameter; the realised step length is 2 sin(alpha_h)."""
    alpha_p = np.arcsin(0.5 * q)
    P = c * alpha_p * np.tan(alpha_p)
    z0 = list(seed) if seed is not None else [alpha_p, -c * alpha_p, (1 - np.cos(2 * alpha_p)) * (-c * alpha_p)]
    z_fp, T, ok = R.find_fixed_point(z0, R.GAM, R.KHIP, P, R.RTOL, R.ATOL, delta=1e-7)
    if not ok:
        return None
    J = R.jacobian_3d(z_fp, R.GAM, R.KHIP, P, R.RTOL, R.ATOL, R.DELTA)
    if J is None:
        return None
    lam_max = float(np.max(np.abs(R.sorted_eigs(J))))
    alpha_h = float(z_fp[0])
    s = 2.0 * np.sin(alpha_h)
    return dict(q=q, alpha_h=alpha_h, s=s, v=s / T, lam_max=lam_max, z=list(z_fp))


REF = solve_at(0.40, 1.04, None)   # converges from the geometric guess; seed for all c


def anchor_for(c):
    """Reach the s=0.40 fixed point at coefficient c by continuation in c from 1.04
    (the geometric initial guess does not converge directly for c != 1.04)."""
    if abs(c - 1.04) < 1e-9:
        return REF
    n = int(round(abs(c - 1.04) / 0.01))
    seed = REF["z"]
    r = REF
    for ct in np.linspace(1.04, c, n + 1)[1:]:
        r = solve_at(0.40, float(ct), seed)
        if r is None:
            raise RuntimeError(f"c-continuation failed near c={ct:.3f}")
        seed = r["z"]
    return r


def trace(c):
    anchor = anchor_for(c)
    rows = [anchor]
    sd = anchor["z"]
    for s in np.arange(0.399, 0.0099, -0.001):
        r = solve_at(float(s), c, sd)
        if r is None:
            break
        sd = r["z"]
        rows.append(r)
    su = anchor["z"]
    for s in np.arange(0.401, 0.801, 0.001):
        r = solve_at(float(s), c, su)
        if r is None:
            break
        su = r["z"]
        rows.append(r)
    rows.sort(key=lambda r: r["q"])
    v = np.array([r["v"] for r in rows])
    qq = np.array([r["q"] for r in rows])
    ah = np.array([r["alpha_h"] for r in rows])
    sr = np.array([r["s"] for r in rows])
    lm = np.array([r["lam_max"] for r in rows])
    nh = -np.log(2.0) / np.log(np.clip(lm, 1e-12, 0.9999999))
    return v, lm, nh, qq, ah, sr


plt.rcParams.update({"font.size": 11})

# --- compute all traces once (expensive), then plot both mono + color ---
traces = []
print(f"{'c':>5} {'n':>4} {'v_lo':>7} {'Nh_lo':>7} {'v~0.016':>8} {'Nh':>6} {'v_hi':>7} {'Nh_hi':>6}")
for c in CS:
    v, lm, nh, qq, ah, sr = trace(c)
    traces.append((c, v, nh, qq, ah, sr))
    i_lo = int(np.argmin(v)); i_mid = int(np.argmin(np.abs(v - 0.016))); i_hi = int(np.argmax(v))
    print(f"{c:5.2f} {len(v):4d} {v[i_lo]:7.4f} {nh[i_lo]:7.1f} {v[i_mid]:8.4f} {nh[i_mid]:6.1f} {v[i_hi]:7.4f} {nh[i_hi]:6.1f}")

# the traced branches, so Supplementary S4 can be replotted without recomputing
import csv as _csv
with open(os.path.join(ROOT, os.environ.get("SLOWWALK_DATA_DIR", "data"), "supp_csweep.csv"), "w", newline="") as _fh:
    _w = _csv.writer(_fh)
    _w.writerow(["c", "q", "alpha_h", "s", "v", "lam_max", "N_half"])
    for c, v, nh, qq, ah, sr in traces:
        lm = np.exp(-np.log(2.0) / nh)
        for vi, li, ni, qi, ai, si in zip(v, lm, nh, qq, ah, sr):
            # q at three decimals is the grid key; the realised s must keep its
            # digits or every branch would collapse onto the same label.
            _w.writerow(["%.2f" % c, "%.3f" % qi, "%.10g" % ai, "%.10g" % si,
                         "%.10g" % vi, "%.10f" % li, "%.4f" % ni])
print("saved: data/supp_csweep.csv")


def make(palette, suffix):
    """palette: list of (color, linestyle, dashes, lw) in CS order."""
    fig, ax = plt.subplots(figsize=(7.4, 4.9))
    for (c, v, nh), (col, ls, dash, lw) in zip(traces, palette):
        lab = f"$c={c:.2f}$" + ("  (main)" if abs(c - 1.04) < 1e-9 else "")
        ax.plot(v, nh, ls, color=col, lw=lw, dashes=dash, label=lab,
                zorder=4 if abs(c - 1.04) < 1e-9 else 3)
    ax.axvline(0.041, color="0.6", ls=":", lw=1.0, zorder=1)
    ax.annotate(r"$N_{1/2}$ rises steeply toward the slow end for every $c$,"
                + "\nwith no fold encountered along the branch traced",
                xy=(0.006, 200), xytext=(0.055, 250), fontsize=9, color="0.3", ha="left",
                arrowprops=dict(arrowstyle="->", lw=0.9, color="0.55"))
    ax.set_yscale("log")
    ax.set_xlabel(r"Dimensionless speed  $v$")
    ax.set_ylabel(r"Perturbation half-life  $N_{1/2}$ (steps)")
    ax.set_xlim(0.0, 0.235)
    ax.set_ylim(1, 1100)
    ax.set_title(r"Push-off coefficient $c$ in $P(s)=c\,\alpha\tan\alpha$: low-speed weakening is robust",
                 loc="left", fontsize=10.5, fontweight="bold")
    ax.legend(fontsize=9.5, frameon=False, loc="upper right")
    ax.spines[["top", "right"]].set_visible(False)
    fig.tight_layout()
    out = OUT.replace(".png", suffix + ".png")
    fig.savefig(out, dpi=300)
    fig.savefig(out.replace(".png", ".tiff"), dpi=300)
    # vector copy for submission: Elsevier asks 1000 dpi for bitmapped line
    # drawings, which vector output makes moot.
    fig.savefig(out.replace(".png", ".pdf"))
    plt.close(fig)
    print("saved:", out)


# grayscale (unchanged)
MONO = [("0.0", "-", (None, None), 1.6), ("0.30", "--", (5, 2), 1.7),
        ("0.0", "-", (None, None), 2.6), ("0.45", ":", (1, 1), 1.8)]
# colorblind-safe (Okabe-Ito): c=0.8 blue, 1.0 green, 1.04 vermillion (bold, main), 1.2 purple
COLOR = [("#0072B2", "-", (None, None), 1.8), ("#009E73", "--", (5, 2), 1.9),
         ("#D55E00", "-", (None, None), 2.6), ("#CC79A7", ":", (1, 1), 2.0)]
make(MONO, "")
make(COLOR, "_color")

File: code_matlab/run_csweep_extended.m
%RUN_CSWEEP_EXTENDED  Push-off-coefficient robustness (port of
%   code/run_csweep_extended.py).  For c in {0.80, 1.00, 1.04, 1.20} the branch
%   with P(s) = c alpha tan(alpha) is traced by csweep_branch.m (seeded
%   continuation over s in [0.01, 0.80], anchored at s = 0.40 by continuation
%   in c from 1.04).  Branches are read from data_matlab/branches/csweep_c<c>.mat
%   and computed here if missing (several minutes each; to parallelise, call
%   csweep_branch(c) in separate MATLAB sessions first).
%   Writes data_matlab/supp_csweep.csv (same schema as data/supp_csweep.csv)
%          figures_matlab/sens_pushoff_c{,_color}.{png,tiff,pdf}
CS = [0.80, 1.00, 1.04, 1.20];
here = fileparts(mfilename('fullpath'));
datadir = fullfile(here, '..', 'data_matlab');
figdir = fullfile(here, '..', 'figures_matlab');
if ~exist(datadir, 'dir'), mkdir(datadir); end
if ~exist(figdir, 'dir'), mkdir(figdir); end
OUT = fullfile(figdir, 'sens_pushoff_c.png');

tic;
traces = cell(1, numel(CS));      % {c, v, nh}
fprintf('%5s %4s %7s %7s %8s %6s %7s %6s\n', 'c', 'n', 'v_lo', 'Nh_lo', 'v~0.016', 'Nh', 'v_hi', 'Nh_hi');
for i = 1:numel(CS)
    f = fullfile(datadir, 'branches', sprintf('csweep_c%.2f.mat', CS(i)));
    if ~exist(f, 'file'), csweep_branch(CS(i)); end
    b = load(f);
    v = b.v(:)'; lm = b.lam_max(:)';
    nh = -log(2.0) ./ log(min(max(lm, 1e-12), 0.9999999));   % np.clip(lm, 1e-12, 0.9999999)
    % b.s is the branch parameter q; alpha_h = Z(1,:) is what the gait realises.
    a_h = b.Z(1, :); s_real = 2.0 * sin(a_h); v_real = s_real ./ b.T(:)';
    traces{i} = {CS(i), v_real, nh, b.s(:)', a_h, s_real};
    v = v_real;
    [~, i_lo] = min(v); [~, i_mid] = min(abs(v - 0.016)); [~, i_hi] = max(v);
    fprintf('%5.2f %4d %7.4f %7.1f %8.4f %6.1f %7.4f %6.1f\n', CS(i), numel(v), ...
            v(i_lo), nh(i_lo), v(i_mid), nh(i_mid), v(i_hi), nh(i_hi));
end

% the traced branches, so Supplementary S4 can be replotted without recomputing
fh = fopen(fullfile(datadir, 'supp_csweep.csv'), 'w');
fprintf(fh, 'c,q,alpha_h,s,v,lam_max,N_half\n');
for i = 1:numel(CS)
    [c, v, nh, qq, a_h, s_real] = traces{i}{:};
    lm = exp(-log(2.0) ./ nh);
    for j = 1:numel(v)
        fprintf(fh, '%.2f,%.3f,%.10g,%.10g,%.10g,%.10f,%.4f\n', ...
                c, qq(j), a_h(j), s_real(j), v(j), lm(j), nh(j));
    end
end
fclose(fh);
fprintf('saved: %s\n', fullfile(datadir, 'supp_csweep.csv'));

% grayscale: {colour, linestyle, linewidth} in CS order
MONO  = {{[0 0 0], '-', 1.6}, {[0.30 0.30 0.30], '--', 1.7}, {[0 0 0], '-', 2.6}, {[0.45 0.45 0.45], ':', 1.8}};
% colorblind-safe (Okabe-Ito): c=0.8 blue, 1.0 green, 1.04 vermillion (bold, main), 1.2 purple
COLOR = {{'#0072B2', '-', 1.8}, {'#009E73', '--', 1.9}, {'#D55E00', '-', 2.6}, {'#CC79A7', ':', 2.0}};
make_figure(traces, MONO, '', OUT);
make_figure(traces, COLOR, '_color', OUT);
fprintf('done (%.1f s)\n', toc);

function make_figure(traces, palette, suffix, OUT)
    fig = figure('Visible', 'off', 'Color', 'w', 'Units', 'inches', 'Position', [1 1 7.4 4.9]);
    set(fig, 'DefaultAxesFontSize', 11, 'DefaultAxesFontName', 'Helvetica', ...
             'DefaultTextFontName', 'Helvetica');
    ax = axes(fig); hold(ax, 'on');
    h = gobjects(1, numel(traces)); labels = cell(1, numel(traces)); i_main = 0;
    for i = 1:numel(traces)
        [c, v, nh] = traces{i}{:};
        [col, ls, lw] = palette{i}{:};
        lab = sprintf('{\\itc} = %.2f', c);
        if abs(c - 1.04) < 1e-9, lab = [lab '  (main)']; i_main = i; end
        h(i) = plot(ax, v, nh, 'LineStyle', ls, 'Color', col, 'LineWidth', lw);
        labels{i} = lab;
    end
    if i_main > 0, uistack(h(i_main), 'top'); end           % zorder 4 for the main branch
    hx = xline(ax, 0.041, ':', 'Color', [0.6 0.6 0.6], 'LineWidth', 1.0);
    uistack(hx, 'bottom');
    set(ax, 'YScale', 'log', 'XLim', [0 0.235], 'YLim', [1 1100], 'Box', 'off', ...
            'TickDir', 'out', 'Layer', 'top');
    ax.XAxis.TickLabelFormat = '%.2f';
    ax.YTick = [1 10 100 1000];
    ax.YMinorTick = 'off';
    xlabel(ax, 'Dimensionless speed  {\itv}', 'Interpreter', 'tex');
    ylabel(ax, 'Perturbation half-life  {\itN}_{1/2} (steps)', 'Interpreter', 'tex');
    title(ax, ['Push-off coefficient {\itc} in {\itP}({\its}) = {\itc}{\it\alpha}tan{\it\alpha}: ' ...
               'low-speed weakening is robust'], 'Interpreter', 'tex', 'FontSize', 10.5, 'FontWeight', 'bold');
    ax.TitleHorizontalAlignment = 'left';
    lg = legend(ax, h, labels, 'Location', 'northeast', 'Box', 'off', 'FontSize', 9.5, 'Interpreter', 'tex');
    lg.ItemTokenSize = [28 18];
    % emulate tight_layout: fill the figure, small outer margin
    ax.PositionConstraint = 'outerposition';
    ax.OuterPosition = [0.005 0.005 0.99 0.99];
    drawnow;
    % annotation: text at (0.055, 250), arrow to (0.006, 200)
    text(ax, 0.055, 250, {'{\itN}_{1/2} rises steeply toward the slow end for every {\itc},', ...
                          'with no fold encountered along the branch traced'}, ...
         'FontSize', 9, 'Color', [0.3 0.3 0.3], 'HorizontalAlignment', 'left', ...
         'VerticalAlignment', 'bottom', 'Interpreter', 'tex');
    p0 = data2fig(ax, 0.055, 250); p1 = data2fig(ax, 0.006, 200);
    annotation(fig, 'arrow', [p0(1) p1(1)], [p0(2) p1(2)], 'Color', [0.55 0.55 0.55], ...
               'LineWidth', 0.9, 'HeadStyle', 'plain', 'HeadLength', 6, 'HeadWidth', 6);
    % export at the figure size (7.4 x 4.9 in): png/tiff at 300 dpi, pdf vector
    set(fig, 'PaperUnits', 'inches', 'PaperSize', [7.4 4.9], 'PaperPosition', [0 0 7.4 4.9], ...
             'InvertHardcopy', 'off');
    out = strrep(OUT, '.png', [suffix '.png']);
    print(fig, out, '-dpng', '-r300');
    print(fig, strrep(out, '.png', '.tiff'), '-dtiff', '-r300');
    print(fig, strrep(out, '.png', '.pdf'), '-dpdf', '-painters');
    close(fig);
    fprintf('saved: %s\n', out);
end

function p = data2fig(ax, x, y)
    % data coordinates -> normalized figure coordinates (log y axis)
    pos = ax.Position; xl = ax.XLim; yl = ax.YLim;
    fx = (x - xl(1)) / (xl(2) - xl(1));
    fy = (log10(y) - log10(yl(1))) / (log10(yl(2)) - log10(yl(1)));
    p = [pos(1) + fx * pos(3), pos(2) + fy * pos(4)];
end
